### PR TITLE
Fix have_tag with blocks

### PR DESCRIPTION
**Pull request by allenwq**
Code inside blocks are actually not been called, so some tests are wrong but we didn't find it.

### CODE REVIEW

**Comment by lowjoel** [comment]
This isn't fixing the problem, you're just avoiding the problem. `has_selector` is Capybara and not the rspec matcher.

**Comment by allenwq** [comment]
Why I do this is because:
1. We should use capybara matchers in feature tests, other than deprecated have_tag matchers
2. I don't think <code>have_tag</code>  can nest inside antoher <code>have_tag</code>

they only support:
```
  have_tag(xxx) do
     with_tag(xxx)
  end
```

**Comment by lowjoel** [comment]
Read the rspec-html-matchers gem documentation. I know it's deprecated in capybara because it was not implemented properly. Here we are using a gem that does proper html validation using nokogiri.

**Comment by allenwq** [comment]
I read their document,  but they never mentioned the <code>have_tag { have_tag }</code> syntax

**Comment by lowjoel** [comment]
Did I mix up have_tag and with_tag?

**Comment by allenwq** [comment]
Our old code is like this right ?
```
expect(page).to have_tag('div.global-announcement') do
  have_tag('div.panel-heading', with_text: announcement.title)
  have_tag('div.panel-body', with_text: announcement.content)
end
```
while the things inside block are not executed.

**Comment by allenwq** [comment]
and <code>with_tag</code> also not works well in this case...

see errors:
```
     Failure/Error: with_text announcement.title
       "Announcement 6" expected within "div.panel-heading" in following template:
       <div class="panel panel-primary global-announcement">
       <div class="panel-heading">
       <div class="pull-right"><button aria-label="Close" class="close" data-dismiss="alert" data-target=".global-announcement" type="button"><span aria-hidden="true">&times;</span></button></div>Announcement 6</div>
       <div class="panel-body">Content 6</div>
       </div>
```

I think it expects the content directly under the <code>div.panel-heading</code>

**Comment by lowjoel** [comment]
Can you push the one using `have_tag`/`with_tag`? `have_tag`/have_tag` is wrong -- I must have read the documentation wrongly.

**Comment by allenwq** [comment]
@lowjoel  Tried for few hours, cannot get with_tag to work.

it does not work if there are anything before the tags you are checking, Like:

Failed case:
``` ruby
expect('<div class="parent"><div class="child"><p>other things</p>Content</div></div>').to have_tag('div.parent') do
  with_tag('div.child', text: 'Content')
end
```

Passd case:
``` ruby
expect('<div class="parent"><div class="child">Content<p>other things</p></div></div>').to have_tag('div.parent') do
  with_tag('div.child', text: 'Content')
end
```

**Comment by allenwq** [comment]
I think we are supposed to use
```
 have_tag('tag', content: 'content')
```
but their content matching is not working and always returns true, seems that all their not working stuffs returns true, so untrustable..

**Comment by lowjoel** [comment]
In your failing test case, you asked for the textual representation of "<p>other things</p>Content" which cannot be "Content" but "other thingsContent".

I think this is more of unfamiliarity with libxml than the gem not working.

**Comment by allenwq** [comment]
It's true that I'm unfamiliar with libxml. but see my second comment, it's also about gem not working.

**Comment by lowjoel** [comment]
Have you debugged into where it's failing? Before declaring something as not working, it is only fair that you try to debug into it.

Nice side effect: you either learn about how to properly use the gem, or otherwise you have a new PR... :) good exercise.

**Comment by allenwq** [comment]
Things like this passed, I haven't further investigated yet:
```
    expect('<div>Content</div>').to have_tag('div', content:'loooll')
```

Well, I'll see if I can find the cause.

**Comment by lowjoel** [comment]
@allenwq there's no `content` key for the hash given to `have_tag`

**Comment by allenwq** [comment]
@lowjoel  they mentioned this in their github home page (https://github.com/kucaahbe/rspec-html-matchers):
```
# mymock.text == 'Some content here'
expect('<p> Some content&nbsphere</p>').to have_tag('p', :content => mymock.text)
# or
expect('<p> Some content&nbsphere</p>').to have_tag('p') do
  with_content mymock.text
end
```

But I just checked their source code
they have something like:
```
    def text_right?
      return true unless @options[:text]
      ....
    end
```
So <code>:content</code> tag seems been ignored.

**Comment by lowjoel** [comment]
I don't see `:content` anywhere in their source file, either, they probably mean `:text`

**Comment by lowjoel** [comment]
@allenwq ...i thought we need to fix this?

**Comment by allenwq** [comment]
Sorry, I occasionally deleted the branch so that it was closed.

File: spec/features/global_announcements_spec.rb [hunk 1]
```diff
@@ -12,21 +12,22 @@
...
+        SystemAnnouncement.destroy_all
         announcement.save!
         visit root_path
       end
 
       it 'shows the announcement' do
         expect(page).to have_tag('div.global-announcement') do
-          have_tag('div.panel-heading', with_text: announcement.title)
-          have_tag('div.panel-body', with_text: announcement.content)
+          with_tag('div.panel-heading', text: format('×%s', announcement.title))
```
Comment (allenwq) [inline spec/features/global_announcements_spec.rb]: why need to add a 'x' before ?

Comment (lowjoel) [inline spec/features/global_announcements_spec.rb]: Because if the definition of text. It is recursive on the subtree.

**Comment by lowjoel** [comment]
@allenwq I can't imagine why the test would fail for Ruby 2.1...

**Comment by allenwq** [comment]
@lowjoel I guess it failed randomly ?

**Comment by lowjoel** [comment]
still need to fix right =.=

**Comment by allenwq** [comment]
Yep, random one more hard ><

**Comment by lowjoel** [comment]
I've run it like 5, 6 times on my computer and I can never get it to show me the same failure...

**Comment by lowjoel** [comment]
@allenwq can you try it on your computer, see if it fails?

**Comment by lowjoel** [comment]
...i know why, I didn't push :(

**Comment by allenwq** [comment]
Thanks ! I just came back :(

**Pull request merged**